### PR TITLE
Fix/get cities names to buy apartment faster

### PR DESCRIPTION
**Pull request by ma28alhams**

### CODE REVIEW

File: src/main/kotlin/interactor/GetCityNamesToBuyApartmentFasterInteractor.kt [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    private fun calculateYearsNeededToBuyApartment(
+        salary: Float,
+        pricePerSquareMeter: Float,
+        squareMeter: Int
+    ): Float {
+        if (squareMeter < 1) throw InvalidMetersException("Square meter can't be less than one")
+        return pricePerSquareMeter * squareMeter / salary / MONTHS_IN_YEAR
+    }
+
+    companion object {
```
Comment (devsadeq) [inline src/main/kotlin/interactor/GetCityNamesToBuyApartmentFasterInteractor.kt:60]: it's a common convention in Kotlin to define the **companion object** at the beginning of the class after the class header.

File: src/test/kotlin/interactor/GetCityNamesToBuyApartmentFasterInteractorTest.kt [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        val squareMeter = 100
+        //When find the cities to buy apartment faster
+        val actual =
+            nullPricesInteractor
+                .execute(
+                    limit = limit,
+                    squareMeter = squareMeter
+                )
+        //Then
+        assertTrue(actual.isEmpty())
```
Comment (devsadeq) [inline src/test/kotlin/interactor/GetCityNamesToBuyApartmentFasterInteractorTest.kt:177]: Bareq says no dots inside assertions

File: src/test/kotlin/interactor/GetCityNamesToBuyApartmentFasterInteractorTest.kt [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    fun setup() {
+        dataSource = FakeDataSource()
+        interactor = GetCityNamesToBuyApartmentFasterInteractor(dataSource)
+        emptyListInteractor = GetCityNamesToBuyApartmentFasterInteractor(FakeDataWithEmptyList())
+        nullPricesInteractor =
+            GetCityNamesToBuyApartmentFasterInteractor(FakeDataForLowQualityData())
+    }
+
+    @Test
+    fun `should return the top 5 cities names that he can buy this apartment faster`() {
```
Comment (devsadeq) [inline src/test/kotlin/interactor/GetCityNamesToBuyApartmentFasterInteractorTest.kt:31]: this name is hard to read ..

try this:
should return top 5 cities names for faster apartment purchase

File: src/main/kotlin/interactor/GetCityNamesToBuyApartmentFasterInteractor.kt [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+class GetCityNamesToBuyApartmentFasterInteractor(
+    private val dataSource: CostOfLivingDataSource
+) {
+    fun execute(
+        limit: Int,
+        squareMeter: Int
+    ): List<Pair<String, Float>> =
+        dataSource.getAllCitiesData()
+            .ifEmpty { throw IllegalStateException("Something went wrong") }
```
Comment (devsadeq) [inline src/main/kotlin/interactor/GetCityNamesToBuyApartmentFasterInteractor.kt:14]: we already have a constants that have different messages that u can use here.

**Comment by devsadeq** [comment]
well done 😍

**Review by devsadeq — APPROVED** [review]

File: src/main/kotlin/interactor/GetCityNamesToBuyApartmentFasterInteractor.kt (hunk 4 above)
Comment (ma28alhams) [reply src/main/kotlin/interactor/GetCityNamesToBuyApartmentFasterInteractor.kt:14]: That's my bad i know, but I'm tired of it

File: src/test/kotlin/interactor/GetCityNamesToBuyApartmentFasterInteractorTest.kt (hunk 3 above)
Comment (ma28alhams) [reply src/test/kotlin/interactor/GetCityNamesToBuyApartmentFasterInteractorTest.kt:31]: I will take yours, but I will change it slightly

File: src/main/kotlin/interactor/GetCityNamesToBuyApartmentFasterInteractor.kt (hunk 1 above)
Comment (ma28alhams) [reply src/main/kotlin/interactor/GetCityNamesToBuyApartmentFasterInteractor.kt:60]: I get it

**Pull request merged**